### ISSUE TITLE
performance

### ISSUE THREAD

**Issue by hannesm** [abeaumont/ocaml-salsa20-core#1]
did you measure performance of this implementation?  was OCaml in the end smart enough to inline the functions in the hot loop?  would be nice to have some mentions of the performance characteristics in the README

**Comment by abeaumont** [comment]
Yes, I have been measuring the performance and is rather bad. I have also looked a bit more closely for the reasons, which I'll try to explain below. I can add them to the README too.

I have used the last test of the testsuite (with 10^6 iterations of Salsa20/20 core function) as a reference. It takes ~6.2s in my laptop. For comparison, same code in Rust takes ~0.2s.

Looking at the generated binary, 2 problems can be easily seen:
* There are array bound checks that could be safely removed.
* There's no inlining of functions.

Using the 4.04.0+flambda compiler both issues are alleviated, and the test takes ~2.7s. There's still some room for improvement, and forcing some inlining (with `[@@inline]`) and eliminating the bound checks (with `-unsafe`) reduces the time to 1.7s.

At this point 3 additional issues become more clear, which are preventing the code to run as fast as it should:
* There's a check to whether the gc should be called in each inner loop block
* There's a call to caml_modify to set the new value, where a register or memory set would be enough
* There's many extra instructions in the basic arithmetic operations because Int32 is used

In other words, each block in the inner loop that could be (taken from the rust code):
```asm
lea ebp, dword [rsi + r9]
rol ebp, 7
xor ebp, r14d
```
is generated as:
```asm
mov rax, qword [r12]
mov rdi, qword [r12 + 0x20]
mov rbx, qword [r12 + 0x40]
movsxd rax, dword [rax + 8]
movsxd rdi, dword [rdi + 8]
add rdi, rax
shl rdi, 0x20
sar rdi, 0x20
sub r15, 0x18
mov rax, obj.caml_young_limit
cmp r15, qword [rax]
jb 0x410d07 ; this is the call to caml_call_gc
lea rsi, qword [r15 + 8]
mov qword [rsi - 8], 0x8ff
mov rax, obj.caml_int32_ops
mov qword [rsi], rax
movsxd rax, dword [rbx + 8]
movabs rbx, 0xffffffff
shl rdi, 0x20
sar rdi, 0x20
mov rdx, rdi
and rdx, rbx
shr rdx, 0x17
shl rdx, 0x20
sar rdx, 0x20
shl rdi, 9
shl rdi, 0x20
sar rdi, 0x20
or rdi, rdx
shl rdi, 0x20
sar rdi, 0x20
xor rdi, rax
mov qword [rsi + 8], rdi
mov rdi, r12
add rdi, 0x40
call sym.caml_modify
```

**Comment by hannesm** [comment]
thanks for the extensive analysis.  `Int32` is rather slow (as described [in its documentation](http://caml.inria.fr/pub/docs/manual-ocaml/libref/Int32.html): Performance notice: values of type int32 occupy more memory space than values of type int, and arithmetic operations on int32 are generally slower than those on int. Use int32 only when the application requires exact 32-bit arithmetic).

My suspicion is that in order to get decent performance, the hot loop should be implemented in C (and similar to nocrypto, avoid to many switches from OCaml to C and back, do allocation on the OCaml side, performance varies depending on the amount of arguments passed (only few can be via registers), use `@@noalloc` annotation).  It is a shame that the inline assembly PR to OCaml https://github.com/ocaml/ocaml/pull/162 (see [this comment](https://github.com/ocaml/ocaml/pull/162#issuecomment-126342195) why) was rejected.

**Comment by hannesm** [comment]
I suspect @dinosaure may have some insight, since he worked on adler32 etc. (IIRC using native OCaml).  maybe he can suggest something here!?

**Comment by dinosaure** [comment]
In Decompress, when I compute the Adler32 checksum, I use a `Int32.t` but only at the beginning and at the end of the hot loop. Indeed, the Adler32 checksum is composed by 2 `int16` (so, we can use two `native int` instead one `Int32.t` and make the `Int32.t` result with these `native int`).

Another things is that, it's better to implement the hot loop in the imperative way (with `ref`, `exception` to jump and imperative loop) and hidden this with a functionnal interface. OCaml optimize this way easily if you constraint the imperative way (with side effect) in one function (otherwise, OCaml can't optimize).

And finally, as @hannesm said, it's better to implement the hot loop in C and hidden in the wild that with an functionnal interface (like the digest hash function in `nocrypto`). A good point is that you can use SSE inside your code and improve the performance (and with `#ifdef ... #endif`, it's easy to maintain this code for all architectures).

PS: about Decompress, this library must be used in the JavaScript code. So, if I want to use a C stub to improve the performance, I need to do the same stub for JavaScript (and this complexify the distribution of Decompress). So, it's like your constraint and what you want.

**Comment by abeaumont** [comment]
Thanks for the feedback. I'll look at both adler32 and nocrypto implementations.

**Comment by abeaumont** [comment]
I rewrote salsa-core function in C, with no allocation, decreasing the test time to 0.3s. Not bad considering that there's 10^6 calls from OCaml to C. Will push the branch soon.

**Comment by abeaumont** [comment]
Branch pushed, changes available at #2. Feedback/reviews are welcome, I'll merge and make a new release afterwards.

**Comment by abeaumont** [comment]
Branch merged, will do a new release soon. [scrypt-kdf](http://github.com/abeaumont/ocaml-scrypt-kdf) test time has been reduced from 107s to 11s on my machine now, not bad :)
